// fpl-api-node/dist/data-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = require("axios");
const NodeCache = require("node-cache");
const errors = require("./errors");
axios_1.default.defaults.baseURL = 'https://fantasy.premierleague.com/drf';
const stdCacheTTL = 1800;
exports.cache = new NodeCache();
function findEntryRoot(entryId) {
    return fetch(`/entry/${entryId}/history`);
}
exports.findEntryRoot = findEntryRoot;
function findEntryEventPicksRoot(entryId, eventNumber) {
    return fetchEvent(`entry/${entryId}/event/${eventNumber}/picks`, eventNumber);
}
exports.findEntryEventPicksRoot = findEntryEventPicksRoot;
function findEntryTransfers(entryId) {
    return fetch(`/entry/${entryId}/transfers`);
}
exports.findEntryTransfers = findEntryTransfers;
function findLiveEvent(eventNumber) {
    return fetchEvent(`/event/${eventNumber}/live`, eventNumber);
}
exports.findLiveEvent = findLiveEvent;
function findLeagueRoot(leagueId, pageNumber = 1) {
    return fetch(`/leagues-classic-standings/${leagueId}?page=${pageNumber}`, false, {
        params: {
            'ls-page': pageNumber,
        },
    });
}
exports.findLeagueRoot = findLeagueRoot;
function getBootstrapData() {
    return fetch('/bootstrap-static');
}
exports.getBootstrapData = getBootstrapData;
function fetchEvent(path, eventNumber) {
    return new Promise((resolve, reject) => {
        const cacheValue = exports.cache.get(path);
        if (cacheValue) {
            resolve(cacheValue);
        }
        else {
            return getBootstrapData().then((data) => {
                const currentEvent = data['current-event'];
                resolve(fetch(path, eventNumber < currentEvent));
            });
        }
    });
}
function fetch(path, cacheForever = false, config = {}) {
    return new Promise((resolve, reject) => {
        const cacheValue = exports.cache.get(path);
        if (cacheValue) {
            resolve(cacheValue);
        }
        else {
            axios_1.default.get(path, config).then((response) => {
                const data = response.data;
                if (Object.keys(data).length > 0 && data.constructor === Object) {
                    exports.cache.set(path, data, cacheForever ? 0 : stdCacheTTL);
                    resolve(data);
                }
                else {
                    if (data.includes('The game is being updated')) {
                        reject(errors.GAME_UPDATING);
                    }
                    else {
                        reject(errors.NOT_FOUND);
                    }
                }
            }).catch(() => {
                reject(errors.NO_RESPONSE);
            });
        }
    });
}
exports.fetch = fetch;
