// mcclowes-scripts/lib/webpack/dev.webpack.config.js
"use strict";

var _coreWebpack = require("./core.webpack.config");

var _coreWebpack2 = _interopRequireDefault(_coreWebpack);

var _webpack = require("webpack");

var _webpack2 = _interopRequireDefault(_webpack);

var _ramda = require("ramda");

var R = _interopRequireWildcard(_ramda);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = R.pipe(R.assoc("devtool", "cheap-eval-source-map"), R.dissocPath(["entry", "vendor"]), R.assoc("devServer", {
	compress: true,
	contentBase: "./build",
	historyApiFallback: true,
	host: "0.0.0.0",
	overlay: true,
	port: _coreWebpack.packageJSON.port || 3000,
	progress: true,
	stats: "minimal"
}), R.assocPath(["output", "filename"], "static/js/[name].bundle.js"), (0, _coreWebpack.addPlugin)(new _webpack2.default.NamedModulesPlugin()))(_coreWebpack2.default);
